src/shader/shaderLightSampling.js

```diff
@@ -27,20 +67,50 @@ LightSampleRec lightsClosestHit( sampler2D lights, uint lightCount, vec3 rayOrig
 			continue;
 		}
 
-		u *= 1.0 / dot(u, u);
-		v *= 1.0 / dot(v, v);
+		u *= 1.0 / dot( u, u );
+		v *= 1.0 / dot( v, v );
 
 		float dist;
-		if ( intersectsRectangle( light.position, normal, u, v, rayOrigin, rayDirection, dist ) ) {
 
-			if ( dist < lightSampleRec.dist || !lightSampleRec.hit ) {
+		if( light.type == 0 ) {
+
+			if ( intersectsRectangle( light.position, normal, u, v, rayOrigin, rayDirection, dist ) ) {
+
+				if ( dist < lightSampleRec.dist || !lightSampleRec.hit ) {
+
+					lightSampleRec.hit = true;
+					lightSampleRec.dist = dist;
+					float cosTheta = dot( rayDirection, normal );
+					lightSampleRec.pdf = ( dist * dist ) / ( light.area * cosTheta );
+					lightSampleRec.emission = light.color * light.intensity;
+					lightSampleRec.direction = rayDirection;
+
+				}
+
+			}
+
+		} else {
+
+			SpotLight spotLight = readSpotLightInfo( lights, l );
+
+			float angle = acos( spotLight.coneCos );
+			float angleTan = tan( angle );
+			float startDistance = spotLight.radius / angleTan;
+
+			vec3 lightPosition = light.position - normal * startDistance;
+			if ( spotLight.lampIntensityScale > 0.0 && spotLight.radius > 0.0 && intersectsCircle( lightPosition, normal, u, v, spotLight.radius * 2.0, rayOrigin, rayDirection, dist ) ) {
```

Comment:
Can you describe what's happening here? Is the spotlight being modeled as a disk here?

Reply:
The lamp itself is being modeled as a circular area light.

Reply:
Is it desirable to have spotlights being modelled as a circular area? I think having them be punctual lights is easier and that is also in line with both glTF and Three.JS. The biggest downside of course is that they won't work in the non MIS case.

Reply:
By default the spot lights are modeled as a single point but optionally you can adjust the "radius" of the spotlight surface (default being 0) to model it as a disc which provides the option for softer shadows. So everything should still match three.js with the option of higher fidelity options that three.js can't provide. In the future this might be a nice way to support point lights / sphere lights with a single implementation.

| radius = 0.0 | radius = 0.5 |
|---|---|
| ![image](https://user-images.githubusercontent.com/734200/175783144-7d685ed6-6926-46f5-8f60-74ffe6e0a0e2.png) | ![image](https://user-images.githubusercontent.com/734200/175783153-55279f09-9eb5-4191-ae5d-d16a1ac5b712.png) |